### ISSUE TITLE
Sync caret position 

### ISSUE THREAD

**Issue by Yaron10** [pnedev/compare-plugin#140]
Hello Pavel,

Thanks again for the new version. Great work.

I still haven't used v3 for a "real Compare session".
I'm posting some issues found in random tests.

STR:
Compare.
Switch to the other file by clicking on its tab.

Result:
The caret position is where it was before Compare.
(A regression in the new version).

Thank you.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> I still haven't used v3 for a "real Compare session".
I'm posting some issues found in random tests.

At that development stage it's perfectly fine, thank you.

> The caret position is where it was before Compare.
(A regression in the new version).

I have done that on purpose actually. Since now it is not necessary for the views to be at the same line (the caret I mean) to be in sync I decided to keep the caret position where it is and let the user set it manually where and when he/she wants to. Instead I'm just changing the visible doc portion when visiting diffs (it's the same as if the user has scrolled to the position himself).
I find it more natural this way. Tell me what you think.

Thanks for commenting and reporting.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> I have done that on purpose actually...
...
(it's the same as if the user has scrolled to the position himself).

I see. Thank you for the explanation.

Other editors include an option to always have the caret visible (i.e. even when mouse-scrolling).
Unfortunately NPP doesn't have this option and your analogy may be correct.

Still, personally I find the old behavior much clearer and the current one somewhat confusing.
Apart from the analogy to mouse-scrolling, do you find the new behavior clearer?
How about a setting?   

(Did you deliberately remove the "Go to first diff after re-Compare" option?).

Thank you very much.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> Apart from the analogy to mouse-scrolling, do you find the new behavior clearer?

To some extent yes. In fact both are OK for me.
Sometimes when I compare I actually want just to check the diffs then close compare and continue at the place I was working before the comparison. In this regard the current implementation is better - while visiting and inspecting the diffs the caret position doesn't change so after closing compare I can just press an arrow key and I'll get back to where I was before.
Another thing I like is that when I want just to view a diff line it is not "blurred" by the line marker because it is not on that line (the caret is not on the line).
Another implication worth mentioning is that when wrap mode is off and you have a horizontal scroll (for a long line) the horizontal scroll is preserved when you are "moving around" (visiting diffs). Now that might be good in some cases and bad in other.

As far as I understand you clearly prefer the caret to be re-positioned, is that correct?

> How about a setting?

Yes, that's possible.

> Did you deliberately remove the "Go to first diff after re-Compare" option?

No, I removed the _Align replacements_ setting but since you haven't deleted your _ComparePlugin.ini_ file from the plugins config folder this is what you are seeing as a side effect. Sorry I didn't mention that. Please delete your _ComparePlugin.ini_and try again.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Good points. Thank you.

I don't use the current-line marker and wrap mode is usually on. :)

> As far as I understand you clearly prefer the caret to be re-positioned, is that correct?

Correct. Besides the clarity I think it should resemble the "Go to..." and "Go to Next/Prev Bookmark" commands.
Thank you for considering a setting.

> No, I removed the Align replacements setting...

Great. Thanks.

BR

**Comment by pnedev** [comment]
Hello Yaron,

OK, I'll add a setting about that.
Thanks.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Great. 
Thank you very much.

BR

**Comment by pnedev** [comment]
Hello Yaron,

I have added user setting for caret position sync but it might not be exactly what you were expecting.
It is difficult to explain but the 'weird' behavior is related to the cases where the caret should be on a blank line (it is truly blank now and the caret cannot be there so it is in the other view). Please try and see for yourself what I'm talking about.

x86 bin: https://ci.appveyor.com/api/buildjobs/udupsxep0ngfq7tf/artifacts/ComparePlugin.dll

You will also need to delete your _Compare.ini_ settings file due to the configuration changes, sorry for the inconvenience.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thank you for adding this setting. I appreciate it.

I did not think about positioning the caret in a blank line.
This indeed makes the issue more complex but I do think it's *highly important*.
(I believe that most CP users would expect the caret to move on navigation).

How about emulating the behavior when the user is trying to position the caret in a blank line?
-- When Next/Prev is a blank line, place the caret *at the end of the previous line* (should it be *next* line in some cases?). 

Blinking that blank line would make the case clearer.

Is it possible to make sure both the blinking blank line and the end of the previous line are visible?

**

Another point:
When *Move Caret On Navigation* is checked, the document should not always scroll up/down (i.e. when Next/Prev is visible, just move the caret).

**

> You will also need to delete your Compare.ini settings file due to the configuration changes, sorry for the inconvenience.

Not a problem at all. :)

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> How about emulating the behavior when the user is trying to position the caret in a blank line?
-- When Next/Prev is a blank line, place the caret at the end of the previous line (should it be next line in some cases?).

Yes, that's the only logical solution.

> Blinking that blank line would make the case clearer.

That is not possible - all blank lines would blink (as mentioned in https://github.com/pnedev/compare-plugin/issues/143#issuecomment-340776738). Let's stick to proper caret positioning only.

> Move Caret On Navigation

I'll change the name - it is indeed clearer, thanks.

> the document should not always scroll up/down (i.e. when Next/Prev is visible, just move the caret).

It is the current behavior, yes. The difference is that when using _Next_ / _Prev_ the caret is positioned on the next **unvisible** diff and not on the next diff. It is perfectly fine solution for the case when the caret is not moved but what about the moving caret case? Perhaps in the latter I need to cycle through the diffs one after the other (disregarding the visibility)?
Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> How about emulating the behavior when the user is trying to position the caret in a blank line?
...
Yes, that's the only logical solution.

Great. Thank you.

If Wrap-mode is off, positioning the caret at the *end* of the previous line might result in horizontal scroll.
Would it be better to place the caret at the *beginning* of the previous line? What do you think?

Another point to consider: placing the caret at the *beginning* of the previous line, it would be impossible to make sure both the blank and the real lines are visible.

> Blinking that blank line would make the case clearer.
...
That is not possible - all blank lines would blink

I forgot that. Sorry.

> Move Caret On Navigation
...
I'll change the name - it is indeed clearer, thanks.

Great. Thank you.

> It is the current behavior, yes...

That's what I meant. The behavior should change according to the setting.

I appreciate your work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> Would it be better to place the caret at the beginning of the previous line? What do you think?

Good point, that's what I intended to do.

> Another point to consider: placing the caret at the beginning of the previous line, it would be impossible to make sure both the blank and the real lines are visible.

You are right, yes. That's why I was planning to first set the caret at the beginning of the previous line and then to scroll the view to the diff line in the other view. The caret will be moved but will be invisible (because it is not possible and the diff is more important). What do you think?

> I forgot that. Sorry.

No problem at all. We are discussing a lot of things - it's normal to forget.

> The behavior should change according to the setting.

You are right, I agree.

Thank you, Yaron.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

The navigation commands are used frequently and I therefore consider the issue quite important.
So I'll allow myself to suggest a more complex solution.

If Wrap-mode is off, place the caret at the beginning of the previous line.  
If Wrap-mode is on, place the caret at the end of the previous line. 

I suppose that in both cases it should be possible to "smart-make-sure" the beginning of the blank block and the previous line (or its end) are visible. Is that correct?
  
NPP allows you to reduce its height and completely hide the editor.
But I'm referring to a normal use.

I have no idea as to the complexity level and the extra work it might require.

Thank you very much. I appreciate your work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> The navigation commands are used frequently and I therefore consider the issue quite important.

Couldn't agree more :+1: 

> If Wrap-mode is off, place the caret at the beginning of the previous line.
If Wrap-mode is on, place the caret at the end of the previous line.

Let me propose another approach.
How about placing the caret in the beginning of the first 'real' line following the blank block in all cases?
This will be consistent regardless of the wrap setting (making it less confusing) and it also readily supports the case where the user copies the missing block (the corresponding diff in the other file) then clicks on the blank block file tab and paste (to have the diff in the blank block file as well).

The line in view will be determined by the diff in the other file - the caret will be placed either above or below the blank block in the current file and then the view will be scrolled so the real diff (in the other file) is visible (centered on the screen). I suppose this answers that question of yours:

> I suppose that in both cases it should be possible to "smart-make-sure" the beginning of the blank block and the previous line (or its end) are visible. Is that correct?

How do you find such approach?
Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'd like to make sure I understand you correctly.

> placing the caret in the beginning of the first 'real' line following the blank block in all cases
...
the caret will be placed either above or below the blank block

I suppose your suggestion is to always place the caret in the line following the blank block.
And in "above or below the blank block" you meant "according to our discussion here".
Is that correct?

> The line in view will be determined by the diff in the other file

So the caret in the active file would not necessarily be visible?

**

The different behavior depending on Wrap-mode is not ideal and your copy/paste point is good indeed.
But IMO the caret *must* be visible (if the setting Move Caret... is checked).
The blinking caret is your guide (or "hot-spot") when using the navigation commands.

If you agree, you'd probably also agree that placing the caret *before* the blank block would be better.

The caret position in the other file would be the real diff, but the visible lines should be determined by the active file.

And my question
> I suppose that in both cases it should be possible to "smart-make-sure" the beginning of the blank block and the previous line (or its end) are visible. Is that correct?

is still valid.

Well, that's my opinion. :)

Thank you very much. I appreciate your patience.
BR

**Comment by pnedev** [comment]
Hello Yaron,

You understand the things perfectly and I agree with everything you say.
The best will be the caret to be visible regardless the 'real' diff in the other file. This also means placing the caret before the blank block.
Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'm glad you agree. I do appreciate it.

BR

**Comment by pnedev** [comment]
Hello Yaron,

The feature is implemented in https://ci.appveyor.com/api/buildjobs/va0m85rcp6ang8o4/artifacts/ComparePlugin.dll (x86).

Thanks.
BR

**Comment by Yaron10** [comment]
Hello Pavel,

Great job indeed! 👍 
Thank you very much. I appreciate your time and work.

STR1:
Check "Move caret on navigation" and "Wrap around diffs".
Go to Last.
Press Next.

Result:
No wrap around (it does work with First/Previous).

**

STR2 (the first post in this thread):
Check "Move caret on navigation".
Compare.
Switch to the other file by clicking on its tab.

Result:
The caret position is where it was before Compare.

BR

**Comment by Yaron10** [comment]
Hello again Pavel,

Regarding STR1:
Could you please test it with [these files](https://github.com/pnedev/compare-plugin/files/1629914/TEST.zip)?

Thank you.
BR

**Comment by pnedev** [comment]
Hello Yaron,

The issues are fixed in this binary (x86):
https://ci.appveyor.com/api/buildjobs/haua2cowwkeor22v/artifacts/ComparePlugin.dll

Thank you for testing and reporting :+1: 
If you find something else please write again.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

The first issue is fixed. 👍 
I do appreciate your work.

I may not have been clear enough regarding the second issue.
Shouldn't the caret position be always synced? 
In v2 the position is synced both on using the navigation commands and when manually moving the caret to a different line.

**

Actually it's more complex in v3.
Let's use the pref files (`prefs.js` is the "New" file):

Compare.
Press Next (the caret is in line 13 before the blank line).

When switching to `prefs Backup.js`, the caret should be in line 14 so that the Next/Prev commands are properly synced.

What do you think? 

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Yes, I thought you meant that but honestly I have no firm opinion what should be the behavior.
You think that v2 approach is the right thing to do?
In v3 it is a bit complicated because in the other view the corresponding blank block might fill the whole screen and the caret will be above it and not visible. That's not much of a problem but might have some unexpected implications. I'll try and we'll see.
Thanks.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I do think the files should be properly synced.

It turns out v3 requires a lot of work and I understand you don't have much free time these days. 
Do you still have the energy? :)

Your v2 is really great and IMO it would be better to postpone the release of the new version until you're in the "right mood".

These lines are based on my personal life experience. Please don't mind that. :)

Anyway, many thanks again.
BR

**Comment by Yaron10** [comment]
Hello Pavel,

Great. :)
And thanks again for your time and patience.

Br

**Comment by pnedev** [comment]
Hello Yaron,

Please give [**this**](https://ci.appveyor.com/api/buildjobs/sqw5ere0ram9q50u/artifacts/ComparePlugin.dll) binary (x86) a try and write back if you find any issues or consider something confusing / not working as expected.

Thank you very much for the feedback!

BR

**Comment by Yaron10** [comment]
Hello Pavel,

That's very good. 👍 
I do appreciate it.

Two fine-tuning comments (please let's use the pref files):

**STR1:**
Compare.
*Manually* place the caret in line 13 in `prefs.js`.
Switch to the other file.

Result:
The caret is line 14. In this case it should be in line 13.
(Would it be complex to implement?).

**STR2:**
Wrap mode is ON.
Line 18 in `prefs.js` is wrapped.
Place the caret in line 20 in `prefs Backup.js`.
Switch to the other file.

Result:
The caret is at *the beginning* of line 18.
For consistency with the navigation behavior it would be nicer if the caret was at *the end* of line 18.

Thanks again for your great work.
BR

**Comment by Yaron10** [comment]
Hello Pavel,

With your permission another related case (pref files):
Compare.
Manually place the caret in line 14 in `prefs.js`.
Press Previous.

Result:
The caret is placed in line 11.
It should be in line 13 (beginning/end depending on Wrap mode).

A similar case: if you place the caret in line 20 in `prefs.js` and press Previous, the caret should be in line 19.

Thank you.
Have a nice weekend.

**Comment by pnedev** [comment]
Hello Yaron,

Thanks for the feedback.

STRs 1 and 2 make sense - those are fixed in this x86 binary:
https://ci.appveyor.com/api/buildjobs/2mfg2b7qatbdnvtr/artifacts/ComparePlugin.dll

The other two are what I consider the expected behavior.
If you go through the diffs from top to bottom of the documents (with _Next_ command) the caret is placed on the first line of each diff block (the top-most part of the diff block).
If on the other hand you try circling through the diffs bottom-up (with _Prev_ command) the caret is placed on the last line of each diff block (the bottom-most part of the diff block).

Try the following using our prefs sample files:
Compare and place manually the caret at line 87 in _prefs_backup_. Go to next diff - notice where the caret is placed.
Now place manually the caret at line 95. Go to prev diff - notice where the caret is placed in that case.
Now do the same with _prefs_ file:
Place the caret at line 63 and go to next.
Place the caret at line 69 and go to prev.

That seems most consistent to me wouldn't you agree?
Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

STR2 is fixed. 👍 
Thank you very much.

I realize I hadn't fully analyzed STR1.
It requires some thinking. :)

Allow me to start with a different question:
Line 32 in 'pref.js` is a real line followed by a blank line.
If the caret is in line 31 and you press Next, you make sure that both line 32 and the following blank line are visible. Is that correct?

But what if you manually (and randomly) place the caret in line 32 and it's *the last visible line*?
If you now press Next, the caret is placed in line 37 and you might miss the correct next diff (i.e. the blank line following 32).

Do you agree this might be a problem?
What do you think about the following solution?
1) When using the navigation commands and the caret "should be in a blank line" but is placed in a real line - blink *the real line* with the blank background color.
2) When the caret is placed in a real line followed by a blank line and you press Next - the default behavior should be leaving the caret in that line (blinking it and making sure the following blank is also visible).
Move the caret to the next diff only if that line has been blinked in the last 30 seconds.

This is complex and the desired behavior in STR1 is related; with your permission, I'll return to STR1 later.
***
>The other two are what I consider the expected behavior.
If you go through the diffs from top to bottom of the documents...

Good point. Allow me to differ. :)
1) The caret should actually be in the blank line. We emulate NPP behavior when the user is manually trying to place the caret there. This is true in both Next and Prev. The previous line is our "replacement-line".
2) Intuitively - it seems as if a diff is skipped. 
3) When the caret is in line 15 and you press Prev, you go to 14; isn't this confusing?
4) The behavior should resemble that of the other file as much as possible.

Thanks again. Highly appreciated.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> But what if you manually (and randomly) place the caret in line 32 and it's the last visible line?
If you now press Next, the caret is placed in line 37 and you might miss the correct next diff (i.e. the blank line following 32).
Do you agree this might be a problem?

That's an interesting corner-case but in that scenario a diff might be missed, yes.

Point 1 seems too much for me. A lot of blinking might get confusing (at least for me - I would wonder what the program is trying to tell me).
Point 2 sounds complex but I got the idea - a functionality to avoid a potential diff miss. :+1: 

Let's move to the navigation commands behavior.

If you have a completely new block of let's say 30 lines in one file you will have a corresponding blank block of 30 lines in the other. Now what is the behavior of CP v2 in case you are under the blank block and call _Prev_? You will get to the last blank line from the block.
CP v3 doesn't have actual blank line and tries to resemble CP v2 by putting the caret just 'before' the blank block which is after it in case the direction is bottom-up. Otherwise it will have to be put before the 30-line blank block. If you do the same steps in the other file (with the newly inserted lines) you will get another behavior - the caret will be at the last line from the block (bottom-up direction with _Prev_ command).

> The previous line is our "replacement-line".

Put in other words the 'previous' line is the first when going down and the last when going up (it's relative to the direction).
If we don't want it to be relative then that should be valid also for the navigation commands behavior in the file without blanks which is:

> The behavior should resemble that of the other file as much as possible.

About 3:

> When the caret is in line 15 and you press Prev, you go to 14; isn't this confusing?

Not at all.
If when going to _Prev_ we go before the blank block and just above that line we have added line and we do again _Prev_ the same thing will happen. Delete line 12 from _prefs_ then compare, place the caret at line 12 in _prefs_ and do _Prev_. Isn't that the same?

Thank you for commenting and for the suggestions.

BR

**Comment by pnedev** [comment]
Hello Yaron,

[**This**](https://ci.appveyor.com/api/buildjobs/7kvkl87iodxaw8ql/artifacts/ComparePlugin.dll) x86 binary fixes the potential problem with navigation diff miss on manual caret setting.

Thank you for catching that :+1: 

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Let's move to the navigation commands behavior.
...

Well argued and explained. 👍 
You're right - the "before" and "after" concept is more important than strictly emulating NPP's behavior on placing the caret in a blank line.
Thank you.

**

I do think that some visual indication to the blank line/block is important.
Is it possible to add a symbol to the beginning/end of a blank block's Compare-Margin?
If it is, how about blinking an arrow pointing to the right at the beginning/end of the blank block?

If it isn't possible, how about blinking a *diagonal arrow* pointing up/down (before/after) in the real line the caret is on? 

Don't you think it would make the navigation clearer?
The user would have to learn about the caret not being in the "true line" anyway.
Adding one more new element won't make it more complex IMO.

***

Your solution to the corner-case is very nice.
But I was thinking that when randomly placing the caret in line 13 and pressing Next - the user might actually want and expect to move to literally-next diff (i.e. the following blank) and not jump some 200 lines down.
In this case - just like pressing Last when the caret is in the last diff - we should let them know: "You're already there".

Obviously, we can't know what the user really wants; but this was the timer idea.

Back to STR1:
The main purpose of syncing the caret position is having the Prev/Next commands synced.
Currently there's some discrepancy in this regard.

If the caret is in line 13 in `pref.js` and you press Next - you go to the *third diff*.
If the caret is in line 13 in `pref.js`, you switch to `prefs Backup.js` and press Next - you go to the *second diff*.

The timer could solve this issue too:
The caret position on switching from line 13 in `pref.js` to `prefs Backup.js` should depend on the timer:
Within 30 seconds the position would be line 14; after 30 seconds the position would be line 13.

Anyway, I understand you don't like it. 
Actually, me neither. :)

Thanks again for your work and time.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for clarifying the timer idea and for the suggestions.
The timer seems unnecessary complex to me, I'm glad you feel that too.

> If the caret is in line 13 in pref.js and you press Next - you go to the third diff.
If the caret is in line 13 in pref.js, you switch to prefs Backup.js and press Next - you go to the second diff.

That was the reason I intentionally implemented the behavior you mentioned [here](https://github.com/pnedev/compare-plugin/issues/140#issuecomment-359116027) - STR1.
But you were right that in case of manual caret re-position it is confusing. Anyway, I don't think that's much of a problem - the user hasn't missed a diff, he/she just needs to hit _Next_ again.

> In this case - just like pressing Last when the caret is in the last diff - we should let them know: "You're already there".

Again, I think it is unnecessary complication. Besides, the blinking in that case will confusingly resemble "You are at the last diff" indication, which is bad.

---------------------

Could you give me a user feedback on something else? I'll appreciate that, thank you.

If you have a selection in one of the files and then switch to the other by clicking into it and move the caret around, the other file's (the one with the selection) caret will not follow (because of the selection).
So far so good.
But if you go to _Next_ / _Prev_ diff in that case the caret in the file with the selection again will not follow and that might bring confusion if you switch back to it by clicking it's tab and then use navigation commands.
Do you think that in the later case the selection should be ignored and the caret should follow again the navigation in the other file?

Thanks for the feedback.

--------------------

Let's move to

> I do think that some visual indication to the blank line/block is important.

I personally don't feel a need for such thing but it won't hurt I suppose. At least we can try and see if that's convenient. :+1: 

> Is it possible to add a symbol to the beginning/end of a blank block's Compare-Margin?

No, unfortunately.

> If it isn't possible, how about blinking a diagonal arrow pointing up/down (before/after) in the real line the caret is on?

It sounds like a good idea :+1: 
Let's try but if it brings unexpected complications we might drop that.
The alignment and caret syncing is currently fairly complex and I still need to fix some regressions (after the last changes we don't go to the first diff after compare) and some synchronization corner cases.

Could you create the corresponding symbols?
Thank you very much for that.

BR

**Comment by pnedev** [comment]
Hello again Yaron,

> ... after the last changes we don't go to the first diff after compare ...

Just to let you know that the regression is fixed **[here](https://ci.appveyor.com/api/buildjobs/2myusnoy558c3ju2/artifacts/ComparePlugin.dll)** (x86).

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Not being able to actually place the caret in a blank line - some minor issues are expected.
Your current implementation seems the best approach. Good analysis and common sense. :)
Thank you.

Thanks also for fixing the last regression.

> If you have a selection in one of the files and then switch to the other by clicking into it...

If the caret is visible, surely this is our hot-spot.
Even if the caret is NOT visible, the user knows (or should know) that in NPP the caret is not always visible. It's like placing the caret in line 1, scrolling down to line 50 and pressing "Next"; - "Next" would be next after line 1.
I therefore think that the current implementation is correct.

Testing it I came across two minor issue related to the "corner-case" and your fix.

1) Placing the caret in line 30 (last invisible) and pressing "Last" - you should jump to the last diff.
2) Placing the caret in line 20, scrolling down to line 70 (line 20 is not visible) and pressing "Next" - you should jump to line 23. 

***

I'm glad you've agreed to try the arrows. I do appreciate it.
[Here](https://github.com/pnedev/compare-plugin/files/1661860/ArrowsUpDown.zip) is the file containing "ArrowUp" (![default](https://user-images.githubusercontent.com/10229320/35359082-6fb14b7e-0161-11e8-9526-fc9d58d7289a.png)) and "ArrowDown" (![default](https://user-images.githubusercontent.com/10229320/35359106-83022680-0161-11e8-8c45-8428e9231a78.png)).
I might play with the images later.

I can think of two minor issues:
1. The blinking symbol might not be seen in a long wrapped line.
2. The diagonal arrows should be flipped if you change the text direction to RTL (not NPP layout; just View -> Text Direction).
It's not worth the extra work. We'll discuss it if you decide to implement it.

I appreciate your work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> Not being able to actually place the caret in a blank line...

> I therefore think that the current implementation is correct.

Thank you for commenting, I appreciate your opinion.

> Testing it I came across two minor issue related to the "corner-case" and your fix.

I couldn't reproduce issue 1. _Last_ should always take you to the last diff and (at least on my side) this is actually the case. Could you give me some more details? Thanks.

Issue 2 is fixed **[here](https://ci.appveyor.com/api/buildjobs/4ikt2h5pqwoaprh7/artifacts/ComparePlugin.dll)**, thanks for catching that :+1: 

Thank you for providing the arrow files.
If I have understood your idea correctly those should blink only when the caret is placed in the line next to the blank block. I therefore imagined that the arrows would be grey. What do you think?
Apart from the color, they look great :+1: 

What do you think about not actually blinking the arrow but showing it for half a second (for example) and hiding it? It will resemble something like 'pointing' to the blank block for a while.

> It's not worth the extra work.

That's also what I think, thanks. If you are concerned with the arrow image gradient then you can simply remove it (make the arrows with even color).

Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> I couldn't reproduce issue 1.

I tested it with the binary you uploaded in https://github.com/pnedev/compare-plugin/issues/140#issuecomment-359828108. It was fixed in the file you uploaded the next day. 👍 
My apologies. And thank you for fixing the two issues.

I got the arrow icons from a free-icons site. They are relatively complex and I couldn't have made them in such a short time. I just played a bit with the color.

> I therefore imagined that the arrows would be grey.

That's a good point. 
OTOH, gray as a foreground color is somewhat "gray" and dull IMO. But it's considered "modern". :)
Your decision. [Here](https://github.com/pnedev/compare-plugin/files/1665752/ArrowsUpDownGray.zip) is the gray version.

Displaying the arrows for 30 seconds can be even better than blinking. 👍 
On using the navigation commands again (and actually moving to a different location), the "pointer" should be removed even within 30 seconds, correct?

The problem with RTL direction is that the Compare-Margin is on the right and the arrows are pointing out...
Something like ![default](https://user-images.githubusercontent.com/10229320/35414122-4236680e-022a-11e8-9e15-9327bf97f37b.png) wouldn't be as clear as the diagonal arrows, right?

Thank you very much. Highly appreciated.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> My apologies. And thank you for fixing the two issues.

No problem, you are welcome. Thanks for catching those.
And thank you for the updated icons. :+1: 

30 seconds? Are you sure?
It can be done of course but isn't it unnecessary complication?

> The problem with RTL direction is...

I got your point, thanks for explaining. I don't think the down-arrow is unclear and it is symmetrical. I like it. I also think that '>' (rotated clockwise to point down and counterclockwise to point up respectively) is fine too. What is more appealing to you?

BR

**Comment by pnedev** [comment]
Hello Yaron,

I've made some fixes to errors I noticed in different navigation modes.
There are still several issues to fix (for example when navigating _Find Unique_ results) but if you are testing meanwhile, please use the **[latest](https://ci.appveyor.com/api/buildjobs/cere4sf8h9efmhl5/artifacts/ComparePlugin.dll)** x86 binary.

Thank you.
BR

**Comment by pnedev** [comment]
Hi Yaron,

Another issue fixed - **[this](https://ci.appveyor.com/api/buildjobs/nt8wpmqd0vjde687/artifacts/ComparePlugin.dll)** is the latest binary. _Find Unique_ results navigation still has problems but other than that navigation should be working fine now (both with and without caret move).

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'm glad you find the time to further improve CP. Great progress.
I appreciate your work.

> 30 seconds? Are you sure?
It can be done of course but isn't it unnecessary complication?

Having several arrows at the same time would be slightly strange IMO.
But more importantly - if you have a single line between two blanks, it can be both "before" and "after".
So when pressing Next and Prev within 30 seconds, you may get two conflicting arrows in the same line.
(You can check and clear an existing arrow in the current line, but I suppose this wouldn't be significantly less complex).

> Something like --- default wouldn't be as clear as the diagonal arrows, right?
...
I don't think the down-arrow is unclear and it is symmetrical...

I meant "clear/clearer" in understanding the symbol's purpose and meaning.
In a not-wrapped line I'm pretty sure the diagonal arrows are better as they are pointing directly to the blank.
I'm not sure about a wrapped line.
Also, the ![default](https://user-images.githubusercontent.com/10229320/35461451-b19029bc-02f0-11e8-962b-9f1eea0faf31.png) icon is often used to "Go to End/Last".

Thanks again.
BR

**Comment by Yaron10** [comment]
Hello Pavel,

This is a chat. :)
Thank you very much. 👍 

Have a nice weekend.

**Comment by pnedev** [comment]
Hello Yaron,

Indeed :)
Thank you.

> Having several arrows at the same time would be slightly strange IMO.

It shouldn't happen for sure but what I meant was the time interval. 30 seconds seem a lot of time to keep the arrow mark. I was thinking of something like a short blink - max 1 second.
What would you say?
The idea is to show the arrow only for lines that are next to the blank block (as we cannot position the caret on a blank itself), correct?

> Also, the ... icon is often used to "Go to End/Last".

You are right. Let's keep the other arrows then and let's also switch them for RTL (I will need a horizontally flipped icons for that).
Thanks.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I also thought 30 seconds was too long but I thought you had suggested it.
For some reason I read your "showing it for half a second (for example)" as "half a minute". Sorry.

I suppose it would be easier for you to set the exact time when actually testing it.

>The idea is to show the arrow only for lines that are next to the blank block (as we cannot position the caret on a blank itself), correct?

Indeed.

> You are right. Let's keep the other arrows then and let's also switch them for RTL (I will need a horizontally flipped icons for that).

Great.
I apologize for the extra work. I didn't think about it initially.
With your permission, an important note: in this case you should check for RTL *Text Direction*; it's not like flipping the Set First icon depending on the entire window layout.

And what do you think about ![default](https://user-images.githubusercontent.com/10229320/35487029-9bad0134-047f-11e8-8ab8-86f02acca973.png) and ![default](https://user-images.githubusercontent.com/10229320/35487034-b53d79a8-047f-11e8-994b-97dc963b459d.png)?
The line is gray and the green arrow adds some color.

Thank you very much. I do appreciate your great work.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> I suppose it would be easier for you to set the exact time when actually testing it.

Not exactly. The implementation depends on the time interval. I can either just wait that time and hide the arrow (in case of short interval because it will block the user input) or I can implement a timer if the interval is longer so the user input is not blocked. I guess I'll have to do the later, thanks.

> I apologize for the extra work. I didn't think about it initially.

That is by no means your fault, no problem. CP is RTL aware (thanks to you) which is great and we should do our best to keep it that way. :+1: 
I'll be checking the text direction, thanks.

The icons with the gray line and the color arrow look great, thanks. :+1: 
Wouldn't it be better to have the arrow in blue rather than green? I would like to avoid any similarities with the green plus sign for added lines (and the red minus in that regard) even if it is just the color.
What do you think?

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Not exactly. The implementation depends on the time interval. I can either just wait..

Thank you for the explanation and for the better (and probably more complex) implementation. 

> CP is RTL aware (thanks to you) which is great and we should do our best to keep it that way.

Highly appreciated. :)

>Wouldn't it be better to have the arrow in blue rather than green? I would like to avoid any similarities with the green plus sign for added lines (and the red minus in that regard) even if it is just the color.

Good point. 
[Here](https://github.com/pnedev/compare-plugin/files/1675644/ArrowsUpDown.zip) is the blue LTR/RTL version ![default](https://user-images.githubusercontent.com/10229320/35539710-6a97a932-055b-11e8-8685-06c2717c3850.png). The file contains also a brown version ![default](https://user-images.githubusercontent.com/10229320/35539761-9b46ab28-055b-11e8-87d1-63118608fda1.png) (better fits the gray? :) ). Whichever you prefer.

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Many thanks for the symbols! :+1: 

BR

**Comment by Yaron10** [comment]
Hello Pavel,

I'm glad you like the icons.
And many thanks for your work.

Br

**Comment by pnedev** [comment]
Hello Yaron,

The blank pointing arrow marks are implemented. I have used the blue ones in resemblance to the blue menu command arrow symbols.
The marks timeout is 1 second.

Here is the corresponding x86 plugin dll:
https://ci.appveyor.com/api/buildjobs/dy66n9qq1u9qdv21/artifacts/ComparePlugin.dll

Could you please test the RTL behavior when you have time? I haven't tried it on my side.
Thank you very much, highly appreciated.

P.S. There are still some corner synchronization issues that need to be handled but other than that I consider the caret move on navigation pretty much finished.
Tell me if you have other suggestions / comments, thanks.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thank you very much for adding this feature. The implementation is beautiful and it makes the navigation much clearer. :+1: 
The LTR/RTL also works like a charm. Special thanks.

Seeing it "in action" for the first time - a one-second-glimpse is too short IMO.  
New users might not grasp the arrows' meaning in this short time and it can cause some confusion.
Reading your previous comments, I'm not sure if 5 seconds would complicate anything. Would it? 
***
The arrows crash NPP on my machine.
I've downloaded the latest portable NPP, installed the last CP (without changing any settings) and had the same result.

STR: 
Compare the pref files (prefs.js is active).
Press Next (the down-arrow appears).
Close NPP.

Result:
Windows crash message.

![default](https://user-images.githubusercontent.com/10229320/36454527-04c7dc7a-16a5-11e8-8a35-1db96b244e41.png)

```
‏‏חתימת הבעיה‏‏חתימת הבעיה:
  שם אירוע הבעיה:	BEX
  שם היישום:	notepad++.exe
  גירסת היישום:	7.5.4.0
  חותמת הזמן של היישום:	5a4985b2
  שם מודול התקלה:	SciLexer.dll_unloaded
  גירסת מודול התקלה:	0.0.0.0
  חותמת הזמן של מודול התקלה:	580d5169
  היסט החריגה:	68b1d67d
  קוד החריגה:	c0000005
  נתוני החריגה:	00000008
  גרסת מערכת ההפעלה:	6.1.7601.2.1.0.256.4
  מזהה אזור:	1037
  מידע נוסף 1:	0a9e
  מידע נוסף 2:	0a9e372d3b4ad19135b953a78882e789
  מידע נוסף 3:	0a9e
  מידע נוסף 4:	0a9e372d3b4ad19135b953a78882e789

קרא את הצהרת הפרטיות שלנו באינטרנט:
  http://go.microsoft.com/fwlink/?linkid=104288&clcid=0x040d

‏‏אם הצהרת הפרטיות באינטרנט אינה זמינה, נא קרא את הצהרת הפרטיות שלנו באופן לא מקוון:
  C:\Windows\system32\he-IL\erofflps.txt
``` 
Sorry for the language. :)
Could you please test it on Windows 7?
***
> P.S. There are still some corner synchronization issues...

I'm curious. :)
I actually haven't had a "real compare-session" for some time.

Thanks again for your great work. Highly appreciated.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for testing and for the quick feedback :+1: 

> I'm not sure if 5 seconds would complicate anything. Would it?

There is no problem at all to change the arrow timeout to any value but 5 sec. seems too much to me. I wouldn't like the arrow to appear as permanent mark (which would happen if the timeout is very big).
What about 2 or 3?

> The arrows crash NPP on my machine.

That's not good, thank you for the catch. I suspect what the problem is.

> I'm curious.

Well, the _Find Unique Lines_ navigation is still problematic. Since the lines are not aligned between files the concept is a bit vague. Should we visit on navigation the unique lines in the current file only or also in the other file? I need to clear things out.

And the most problematic aspect of the whole alignment-blanks implementation - the synchronization issues when Notepad++ changes the views line number margin width when the wrap around is enabled (it happens when you have a long file and go from line number with X digits to line number with X+/-1 digits). This changes the lines wrap 'amount' (since the lines margin portion takes different portion of the screen which effectively changes the document width) and triggers realignment that on the other hand also changes the lines appearance, alters the location view (current first visible line, scroll, etc.) and occasionally may lead to new change in the line number margin that starts the whole process again. I've even seen oscillation caused by this event loop.

BR

**Comment by pnedev** [comment]
Hello again Yaron,

Please try if the crash is fixed in this version:
https://ci.appveyor.com/api/buildjobs/72ijp8h0ksrjjt2k/artifacts/ComparePlugin.dll

Timeout set to 3 sec.
Thank you.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Thank you for the quick crash-fix. Great! :+1: 
3 seconds is very good. You're right, 5 seconds would be too long.
And again, brilliant work. I do think it makes the navigation much clearer. Also - in general it makes CP more "sleek".

Allow me one more question:
Is placing the arrows in the bookmarks-margin intentional?
- They belong to CP.
- The bookmarks-margin can be hidden.
- You might get both a bookmark and an arrow in the same line.
***
> Well, the Find Unique Lines navigation is still problematic...

Thanks for elaborating. I appreciate it.
I'm very tired and I find it hard to concentrate now. I'll think about it tomorrow.

BR

**Comment by pnedev** [comment]
Hello Yaron,

> Is placing the arrows in the bookmarks-margin intentional?

Not at all, I haven't even noticed that (what's wrong with me? :)).
Thank you for the excellent catch :+1: 

Here is the modified x86 binary:
https://ci.appveyor.com/api/buildjobs/kbv9i5u1320oajbs/artifacts/ComparePlugin.dll

BR

**Comment by Yaron10** [comment]
Hello Pavel,

There's nothing wrong with you. Quite the contrary. :)
I didn't notice it on the first day either.

Thank you for fixing it.
***
> Well, the Find Unique Lines navigation is still problematic...

Theoretically you can display the unique-lines without even synchronizing the vertical scroll and the caret position (and each file should have an independent navigation).

But I think it would be better to regard the diffs in both files as if they exist in one file (overlapping).
Practically it means "zigzagging" between the diffs in both files but remaining in the active file.

Let's please use the prefs files (prefs.js is active).
The caret is line 40.
Pressing Next moves the caret to line 42.
Pressing Next again moves the caret to line 53 (no unique-line in prefs.js, but line 53 is "aligned" with the unique-line 56 in the other file).
This is the current behavior when using those lines. It should be implemented throughout the navigation.
What do you think? 

BTW, not having a shortcut for "Clear All Compares", how about removing the shortcuts for the "Find Unique" commands? - It might make the menu slightly clearer.
***
> And the most problematic aspect of the whole alignment-blanks implementation...

Reading the following lines, please remember that I have no clue as to the actual implementation. 
This is a question I've been wanting to ask you regardless of this issue:
Would it be possible *not* to insert blanks during the comparison process but rather do it dynamically?
IOW: regard the visible lines as the *only block* you have to align the diffs in, and do it dynamically on scrolling, re-sizing etc..

This would solve the line-number margin-width issue and also might prevent the necessity of inserting long blank blocks. 

Does it even start to make sense? 

If not, can you control the line-number margin-width for the Compare session?

Thanks again for your work. I do appreciate it.
BR

**Comment by pnedev** [comment]
Hello Yaron,

> But I think it would be better to regard the diffs in both files as if they exist in one file (overlapping).

That's also what I think and this is how it should be working now but there are some things to fix there (when line wrap is involved and there are aligned unique lines).

> Would it be possible not to insert blanks during the comparison process but rather do it dynamically?

That's an interesting thought. I cannot think of all the hidden implications upfront but I can try and we'll see how things work. 
What I don't like about this approach is the 'glitchy' appearing behavior. When the user navigates between diffs the screen will always kind-of flicker when the alignment happens and it seems it will be quite often. The things get even worse because the user might use the scroll to 'smoothly' change the viewed lines (not by navigation commands jump) and that will not be smooth at all. Imagine a line that just comes into view and suddenly something happens, some blank portion appears of nowhere and the line is gone who knows where. This things happen now only when the line margin changes or when you resize the window.
The other thing that comes to my mind is that the context (the diff surrounding lines) will be more 'unaligned' (far from each other). If you compare long files the misalignment due to wrapping tends to accumulate going down. If we align only the currently visible section and we are in the lower part of the file then we'll have to compensate a bigger misalignment that will probably lead to having a full blank section above the diff in one of the files.
Anyway, it doesn't hurt to try, that's a good suggestion, thanks :+1: 

> If not, can you control the line-number margin-width for the Compare session?

I think that's not good approach. Notepad++ controls it dynamically (that's why it changes the margin width constantly) and competing with it for control is a bad thing to do.

I'll close this thread now. We'll track any future problems in separate threads.

Thank you for all the testing, feedback and suggestions. I do appreciate it.

BR

**Comment by pnedev** [comment]
Ou, sorry.

I forgot to comment on that:

> BTW, not having a shortcut for "Clear All Compares", how about removing the shortcuts for the "Find Unique" commands? - It might make the menu slightly clearer.

I prefer to keep them. Shortcuts make things more accessible and _Find Unique_ ones don't interfere with anything (AFAIK). Besides, IMO _Clear All Compares_ will seldom be used by the average user (IOW it is not part of the most common compare workflow).
I hope you don't mind.

Thank you once again.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

>That's also what I think and this is how it should be working now but there are some things to fix there (when line wrap is involved and there are aligned unique lines).

Great. Thank you.

After posting yesterday I thought about the following case:
The active file is shorter and there's a unique-line at the end of the other longer file.
What should happen when you reach the last unique-line in the active file and press Next?

How about actually "zigzagging" and switching between the files?
And the concept is: we regard both files as one, and "Next" should be the next NOT-overlapped unique-line *in either file*.
Overlapped unique-lines are considered seen/visited; as is your concept in not moving the caret on navigation.

The active file on pressing "Find Unique" would determine which file is "old/new", but if a unique-line appears first in the non-active file - the focus and caret should be moved to that file.

Let's use the prefs files (prefs.js is active).
The caret is in line 15.
Press Next.
The caret should move to line 21 in the other file.

If this is too complex, I do think that navigating completely separately in each file would be fine; I'm not sure there's a point in synchronizing the vertical scroll if you prefer this approach. 

**

I suggested to remove the long shortcuts knowing your preference not to include shortcuts in general.
That's absolutely fine. Thanks.
***    
> What I don't like about this approach is the 'glitchy' appearing behavior...

I certainly had this in mind and it was a major part in the "Does it even start to make sense?" question.
I was thinking about a second (or so) delay before starting the alignment (so that it wouldn't be too "flickery" on fast scrolling). 

Thank you very much for considering it and for the explanation. I appreciate it.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

> And this is a good approach to the problem also.

I'm glad you agree. It makes it easier to "digest" this slightly tricky approach.
Thank you very much and best of luck.

A minor point:
If the files do not contain any unique-lines, they don't necessarily match (moved or empty lines).
Wouldn't it better to use "Files A and B do not contain unique lines."?

> When I fix the Find Unique navigation problems I plan to move on to Moved-multiple detection behavior and implementation.

Great. I'll be looking forward to resuming this discussion. Thank you.

> I'll try the alignment-of-visible-portion approach some time in the future.

Thank you for that too. I assume it's not an easy task.

BR

**Comment by pnedev** [comment]
Hello Yaron,

> A minor point...

Great, thank you :+1: 

It is fixed here:
https://ci.appveyor.com/api/buildjobs/wjn51t7wun060u5a/artifacts/ComparePlugin.dll

BR

**Comment by Yaron10** [comment]
Hello Pavel,

:+1: 
Thank you. I appreciate it.

BR

**Comment by pnedev** [comment]
Hello Yaron,

'Long time no write' from my side ;)
Hope you are doing well.

I'd just like to let you know that the _Find Unique_ navigation is ready.
You can find the 32 bit binary here:
https://ci.appveyor.com/api/buildjobs/ph65jd403etm7u2p/artifacts/ComparePlugin.dll

If you find any issues with the navigation and caret syncing in the future, please feel free to write (appreciated as always) :+1: 

I'm moving to multiple-move topic.

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Glad you're back. :) Whenever you have free time it's appreciated.
I'm fine, thank you. I hope you are doing well too.

Your implementation is slightly different than what I had in mind; but it actually seems to be much better.
On Next you're rolling the ball (i.e. the caret) down to the first lower block, and on Prev you're pushing it up to the first higher end-of-block.
The behavior depends on Wrap-mode ON/OFF.

This is beautiful and clever; I was really impressed. :+1: :+1: 
Thank you very much. Highly appreciated.

I came across an interesting case.
STR:
Set NPP to `width="960" height="732"`.
Wrap is OFF.
Open the pref files.
`prefsFF.js` is active.
Find Unique.
Place the caret in line 21 in `prefsFFbackup.js`.
Turn Wrap ON.
Press Next.

Result:
Line 21 in `prefsFF.js` blinks.
***
I've come across a minor caret-issue. I'll post later again.
***
Could you please have a look at [this topic](https://notepad-plus-plus.org/community/topic/15438/multiple-errors-building-npp-7-5-6)?
Would I be able to build CP if I upgrade to VS 2017?

Thanks again.
BR

**Comment by pnedev** [comment]
Hello Yaron,

Thank you for the prompt reply (as always) :)

I could'n reproduce the case you described but I'm not sure my resolution was correct.
Could you send me a screenshot exactly before _Press Next_ step (to see the files carets and diff blocks relative positions)? Much appreciated.

Honestly, I don't have any experience with VS2017. Reading the N++ topic it is needed because the new compiler supports the new C++ language features and constructs (C++ 11 and C++ 14 updates that are used in the code). Since the new VS2017 compiler is backwards compatible you shouldn't experience any problems compiling CP.

Thanks and
BR

**Comment by Yaron10** [comment]
Hello Pavel,

> Thank you for the prompt reply...

My pleasure. Thanks again for your excellent work.

> I couldn't reproduce the case...

STR:
Open NPP config.xml and set `width="960" height="732"`.
Wrap is OFF.
Open the pref files.
prefsFF.js is active.
Find Unique.
Place the caret in line 21 in prefsFFbackup.js.

![default](https://user-images.githubusercontent.com/10229320/37728878-1993a96c-2d44-11e8-83b2-17c5365069a5.png)

Turn Wrap ON.
Press Next.

![default](https://user-images.githubusercontent.com/10229320/37728913-342d7dac-2d44-11e8-9ff2-0ec5121dfb68.png)

Result:
Line 21 in prefsFF.js blinks.
***
> Since the new VS2017 compiler is backwards compatible you shouldn't experience any problems compiling CP.

Great. Thanks for that too. :)

BR

**Comment by pnedev** [comment]
Hello Yaron,

I see the problem, thanks for the screenshots.
That is a good catch, great job, thanks :+1: 

The issue should be fixed **[here](https://ci.appveyor.com/api/buildjobs/s85uvf6b3073emvw/artifacts/ComparePlugin.dll)** (x86).

BR

**Comment by Yaron10** [comment]
Hello Pavel,

Great. :+1: 
Thank you very much. Much appreciated.

BR
